src/org/opensolaris/opengrok/configuration/messages/MessageListener.java

```diff
@@ -0,0 +1,503 @@
+
+    private final Gson gson = new Gson();
+
+    private final Map<Class<? extends Message>, List<MessageHandler>> messageHandlers = new HashMap<>();
+
+    private ServerSocket configServerSocket;
+
+    /*
+    initial capacity - default 16
+    initial load factor - default 0.75f
+    initial concurrency level - number of concurrently updating threads (default 16)
+        - just two (the timer, configuration listener) so set it to small value
+    */
+    private final ConcurrentMap<String, SortedSet<AcceptedMessage>> tagMessages = new ConcurrentHashMap<>(16, 0.75f, 5);
+
+    private int messagesInTheSystem = 0;
+
+    private int messageLimit = DEFAULT_MESSAGE_LIMIT;
+
+    private Timer expirationTimer;
+
+    public void addMessageHandler(final Class<? extends Message> messageType, final MessageHandler handler) {
+        if (messageType == null) {
+            throw new IllegalArgumentException("Cannot add handler for null message type");
+        }
+        if (handler == null) {
+            throw new IllegalArgumentException("Cannot add null handler");
+        }
+
+        LOGGER.log(Level.FINEST, "Adding message handler {0} for message type: {1}",
+                new Object[] {handler, messageType});
+
+        synchronized (messageHandlers) {
+            messageHandlers.computeIfAbsent(messageType, key -> new LinkedList<>()).add(handler);
+        }
+    }
+
+    public void removeMessageHandler(final Class<? extends Message> messageType, final MessageHandler handler) {
+        if (messageType == null) {
+            throw new IllegalArgumentException("Cannot remove handler for null message type");
+        }
+        if (handler == null) {
+            throw new IllegalArgumentException("Cannot remove null handler");
+        }
+
+        LOGGER.log(Level.FINEST, "Removing message handler {0} for message type: {1}",
+                new Object[] {handler, messageType});
+
+        synchronized (messageHandlers) {
+            List<MessageHandler> handlers = messageHandlers.get(messageType);
+            if (handlers == null || !handlers.contains(handler)) {
+                LOGGER.log(Level.WARNING, "{0} is not registered as a handler", handler);
+                return;
+            }
+            handlers.remove(handler);
+            if (handlers.isEmpty()) {
+                messageHandlers.remove(messageType);
+            }
+        }
+    }
+
+    /**
+     * Try to stop the configuration listener thread
+     */
+    public void stopListenerThread() {
+        IOUtils.close(configServerSocket);
+    }
+
+    /**
+     * Start a thread to listen on a socket to receive new messages.
+     * The messages can contain various commands for the webapp, including
+     * upload of new configuration.
+     * @param endpoint The socket address to listen on
+     * @throws IOException if the endpoint is not available
+     */
+    public void start(SocketAddress endpoint) throws IOException {
```

Comment:
`startListenerThread` ?

Reply:
`Listener` is already in class name. That it starts different thread or process or something else should not be visible outside the class IMHO. To assure consistency I would rather rename `stopListenerThread` to just `stop`. However, if you prefer `startListenerThread`, I will add it there.